src/perf_tools/mem_prof.cr

```diff
@@ -326,6 +344,144 @@ module PerfTools::MemProf
     end
   end
 
+  # Logs the objects that are transitively reachable to a given *type* of objects,
+  # outputing a mermaid graph to the given *io*. The graph contains an indication of
+  # the field that links an object, or its index if the field is not known.
+  #
+  # Example:
+  #
+  # ```
+  # class A
+  #   @bs = Array(B).new
+  #
+  #   def add(b : B)
+  #     @bs << b
+  #   end
+  # end
+  #
+  # class B
+  #   @c : C
+  #
+  #   def initialize(@c : C)
+  #   end
+  # end
+  #
+  # class C; end
+  #
+  # a = A.new
+  # a.add(B.new C.new)
+  # a.add(B.new C.new)
+  #
+  # PerfTools::MemProf.pretty_log_object_graph STDOUT, C
+  # ```
+  #
+  # produces the following graph:
+  #
+  # ```mermaid
+  # graph LR
+  #   0x109736e80["0x109736e80 C"]
+  #   0x109736e70["0x109736e70 C"]
+  #   0x10972fe40["0x10972fe40 B"] --@c,0--> 0x109736e80
+  #   0x10972fe20["0x10972fe20 (class 0)"] --@(field 0),1--> 0x10972fe40
+  #   0x10972fe60["0x10972fe60 Array(B)"] --@(field 16),2--> 0x10972fe20
+  #   0x10972fe80["0x10972fe80 A"] --@bs,3--> 0x10972fe60
+  #   0x10972fe00["0x10972fe00 B"] --@c,0--> 0x109736e70
+  #   0x10972fe20["0x10972fe20 (class 0)"] --@(field 8),1--> 0x10972fe00
+  # ```
+  #
+  # `(class 0)` is the buffer of the array. The number next to the field is the level of indirection,
+  # counting as 0 for the objects that direct links to an object of the given *type*, 1 for the objects
+  # that have a direct link to an object of the given *type*, and so on.
+  #
+  # The graph is limited to `REF_LIMIT` objects and `REF_LEVEL` levels of indirection.
+  def self.pretty_log_object_graph(io : IO, type : T.class) : Nil forall T
+    GC.collect
+    stopping do
+      type_id = T.crystal_instance_type_id
+      alloc_infos = self.alloc_infos
+      pointers = alloc_infos
+        .select { |_, info| info.type_id == type_id }
+        .map { |ptr, _| {ptr, 0} }
+
+      pointers = REF_LIMIT == 0 ? pointers.to_h : pointers.first(REF_LIMIT).to_h
+      original_pointers = pointers.dup.map { |ptr, _| ptr }
```

Comment:
Functional-like programming is very nice, but we might be manipulating lots of data in this method —I guess? Maybe we can inline the manipulation and avoid intermediary representations?

```crystal
pointers = {} of Pointer => Int32

alloc_infos.each_value do |info|
  next unless info.type_id == type_id
  pointers[ptr] = 0
  break if REF_LIMIT > 0 && pointers.size >= REF_LIMIT
end
```

Also, no need to `dup` + nice method:

```crystal
original_pointers = pointers.keys
```